add documentation for group by directive (#296)
neo4j/docs-graphql: modules/ROOT/pages/directives/schema-configuration/field-configuration.adoc

----

As shown by the previous inputs fields, the `actors` field is not available for filtering on both value and aggregation filters.

== `@groupBy`

This directive defines fields that can be used to group query results.

=== Definition

[source, graphql, indent=0]
----
"""Enables groupBy operations for this field."""
directive @groupBy() on FIELD_DEFINITION
----

=== Usage

> With this definition:

Reviewer: ```suggestion
Consider the following definition:
```

[source, graphql, indent=0]
----
type Movie @node {
    title: String!
    year: Int! @groupBy
}
----

> Now the field `groupBy` exists in the `movieConnection` operation. This fields allow to query movies grouped by the `year` field.

Reviewer: ```suggestion
It adds the field `groupBy` to the `movieConnection` operation.
The field allows you to query movies grouped by the `year` field.
```

> The groupBy field, has one argument `fields`, that defines the field to group by. In this case `year` is the only option

Reviewer: ```suggestion
The `groupBy` field has one argument `fields`, that defines the field to group by.
In this case `year` is the only option.
```

[source, graphql, indent=0]
----
query {
  moviesConnection {
    groupBy(fields: { year: true }) {
      edges {
        node {
          title
          year
        }
      }
    }
  }
}
----

> The `groupBy` field will return an list of results, each result will contain a list of `edges`, with all the results grouped by the `year` field. For example:

Reviewer: ```suggestion
The `groupBy` field returns a list of results and each result contains a list of `edges`, with all results grouped by the `year` field.
For example:
```

[source, json, indent=0]
----
{
    "moviesConnection": {
        "groupBy": [
            {
                "edges": [
                    {
                        "node": {
                            "title": "The Matrix",
                            "year": 1999
                        }
                    },
                    {
                        "title": "Fight Club",
                        "year": 1999
                    }
                ]
            },
            {
                "edges": [
                    {
                        "node": {
                            "title": "Gladiator",
                            "year": 2000
                        }
                    }
                ]
            }
        ]
    }
}
----

> Note that the groupedBy field `year` will be repeated on every node of the edges from each `groupBy` result.

Reviewer: ```suggestion
Note that the `year` field is repeated on every node of the `edges` from each `groupBy` result.
```

Nodes can be grouped by multiple fields:

[source, graphql, indent=0]
----
query {
  moviesConnection {
    groupBy(fields: { year: true, rating: true }) {
      edges {
        node {
          title
          year
        }
      }
    }
  }
}
----

> In this case, all movies with the same `year` and `rating` will be grouped together in the same result.

Reviewer: ```suggestion
In this case, all movies with the same `year` and `rating` are grouped together in the same result.
```

----

As shown by the previous inputs fields, the `actors` field is not available for filtering on both value and aggregation filters.

== `@groupBy`

This directive defines fields that can be used to group query results.

=== Definition

[source, graphql, indent=0]
----
"""Enables groupBy operations for this field."""

> directive @groupBy() on FIELD_DEFINITION

Reviewer 2: should this have the empty paranthesis?

Author: ```suggestion
directive @groupBy on FIELD_DEFINITION
```

----

=== Usage

Consider the following definition:

[source, graphql, indent=0]
----
type Movie @node {
    title: String!
    year: Int! @groupBy
}
----

It adds the field `groupBy` to the `movieConnection` operation.

> The field allows you to query movies grouped by the `year` field.

Reviewer 2: ```suggestion

```

> The `groupBy` field has one argument `fields`, that defines the field to group by.

Reviewer 2: ```suggestion
The `groupBy` field has a single, mandatory argument `fields` which defines the field to group the results by.
The type of the `groupBy` field is a list of results compliant to the Connection Spec.
```

> In this case `year` is the only option.

Reviewer 2: ```suggestion
The following example shows how to query movies grouped by the `year` field.
In this case, `year` is the only field the results can be group by.
```

[source, graphql, indent=0]
----
query {
  moviesConnection {
    groupBy(fields: { year: true }) {
      edges {
        node {
          title
          year
        }
      }
    }
  }
}
----

> The `groupBy` field returns a list of results and each result contains a list of `edges`, with all results grouped by the `year` field.

Reviewer 2: ```suggestion
The result of this query is a list of `edges` wherein all results share the same value of the `year` field.
```

> For example:

Reviewer 2: ```suggestion

```

[source, json, indent=0]
----
{
    "moviesConnection": {
        "groupBy": [
            {
                "edges": [
                    {
                        "node": {
                            "title": "The Matrix",
                            "year": 1999
                        }
                    },
                    {
                        "title": "Fight Club",
                        "year": 1999
                    }
                ]
            },
            {
                "edges": [
                    {
                        "node": {
                            "title": "Gladiator",
                            "year": 2000
                        }
                    }
                ]
            }
        ]
    }
}
----

> Note that the `year` field is repeated on every node of the `edges` from each `groupBy` result.

Reviewer 2: ```suggestion

```

> Nodes can be grouped by multiple fields:

Reviewer 2: ```suggestion
Nodes can be grouped by multiple fields.

Consider the following type definition:

[source, graphql, indent=0]
----
type Movie @node {
    title: String!
    year: Int! @groupBy
    rating: Int! @groupBy
}
----

You can now use both `year` and `rating` fields in the argument of the `groupBy` field.
```
